### DISCUSSION TITLE
ChatGPT+ for automated Snowflake TF resource migration

### DISCUSSION THREAD

**Discussion by chrisweis** [Snowflake-Labs/terraform-provider-snowflake#2609] [Ideas]
Quick tip: I was able to use ChatGPT+'s GPT feature to upload the markdown resource documentation along with a text file that guides the migration process and so far it seems to work pretty well.  I unfortunately wasn't able to share the GPT since it apparently contains copyrighted info (probably "Snowflake").

**Here are the steps** (I believe it requires either ChatGPT+ or Enterprise versions):

1. Get the markdown files to your computer: `git clone https://github.com/Snowflake-Labs/terraform-provider-snowflake`
2. Merge the resources MD files into one file.  (See PowerShell script below)
3. Create a TXT file that guides the conversion (see **Index of Resource Migration.txt** below)
4. Create a GPT using the configuration prompt below.
5. Upload all the files, including the file `technical-documentation/resource_migration.md`
6. Chat with it to get the migrated TF code!

**Here's an example of how it behaves:**
![image](https://github.com/Snowflake-Labs/terraform-provider-snowflake/assets/13931731/dd0eca1a-e39f-4da5-8b6f-62ee917ba551)

**PowerShell command (run in the docs/resources directory):**
```
Get-ChildItem -Filter *.md | Where-Object { $_.Name -ne 'combined.md' } | ForEach-Object {
    "Contents of $($_.Name):"
    "`n"
    Get-Content $_.FullName
    "`n"
} | Set-Content combined.md

```

**Index of Resource Migration.txt**
(Note: This file might not even be necessary)
```
snowflake_database_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_schema_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_table_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_view_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_warehouse_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_tag_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_user_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_stage_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_procedure_grant is migrating to snowflake_grant_privileges_to_account_role
snowflake_account_grant is migrating to snowflake_grant_privileges_to_account_role

snowflake_role_grants is migrating to snowflake_grant_account_role
```

**GPT configuration prompt:**
```
Make a converter for the Snowflake Terraform provider called "Snowflake Terraformer" that converts Hashicorp Configuration Language (HCL) code from deprecated resources to the new resources.  It is driven by uploaded markdown documentation files from the Snowflake Terraform provider in Github (https://github.com/Snowflake-Labs/terraform-provider-snowflake/tree/main/docs) and an index of which resources are being replaced by others.  The logo colors should include purple to match Terraform's color.

It should prioritize matching the generated code to function properly in the new resources, and accurately matching the patterns in the documentation so that they are valid and function properly when run with "terraform validate" or "terraform apply".
```

Also, check out this cool comment/gist if you want to convert DDL to TF: https://github.com/Snowflake-Labs/terraform-provider-snowflake/discussions/685#discussioncomment-5541408

**Comment by gthomson31** [comment]
Will be giving this a try
